---
title: "Project_Bioinformatic resources"
submitted by: "Surya Hembrom, matriculation nr: 229578"
output:
  html_document: default
  pdf_document: default
date: "2023-07-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
```
```{r}

# dataset : lung adenocarcinoma
load("Lung_adenocarcinoma.RData") 

```


```{r}
library("biomaRt")
# get the dataset for Homo sapiens from ensembl
#ensembl <- useMart(biomart="ensembl",dataset="hsapiens_gene_ensembl", host = "uswest")
ensembl <- useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl", mirror = "useast")
# get the filters in dataset
filters <- listFilters(ensembl)
head(filters)
dim(filters)
```
2. Update raw_count_df and r_anno_df extracting only protein coding genes.
a. Use biomaRt package to retrieve the needed information
b. Next tasks should use the new data-frames you have created 


```{r}
# extract all the genes from BiomaRt
human_genes <- getBM(attributes=c("ensembl_gene_id",
                        "external_gene_name",
                              "gene_biotype",
                              "transcript_count",
                              "start_position",
                              "end_position",
                              "chromosome_name",
                              "strand",
                              "description",
                              "version"),
                filters=c("chromosome_name"),
                values=c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22),
                mart = ensembl)
# fetch the protein coding genes
human_genes_protein_coding <- subset(human_genes, human_genes$gene_biotype == "protein_coding")



```

```{r}
# add the  protein coding genes 
### get the list of protein coding genes 
protein_coding_genes_list <- human_genes_protein_coding$ensembl_gene_id

# subset the the annotation df based on protein coding genes list 
filt_r_anno_df <- subset(r_anno_df, r_anno_df$ensembl_gene_id %in% protein_coding_genes_list)
# subset the raw counts df based on protein coding genes list 
raw_counts_df <- as.data.frame(raw_counts_df)
filt_raw_counts_df <- subset(raw_counts_df, rownames(raw_counts_df) %in% protein_coding_genes_list)

```
3. Perform differential expression analysis using edgeR package and select up- and down-regulated genes using a p-value cutoff of 0.01, a log fold change ratio >1.5 forup-regulated genes and < (-1.5) for down-regulated genes and a log CPM >1. Relax the thresholds if no or few results are available. 
a. Use the workflow we developed during the course
b. Filter raw counts data retaining only genes with a raw count >20 in at least 5 Cases or 5 Control samples
c. Create a volcano plot of your results
d. Create an annotated heatmap focusing only on up- and downregulated genes

```{r}
reads_20 <- 20 # min reads for genes 
case_5 <-  5 # number of cases for genes with min 20 reads

# filter genes with raw counts reads > 20
filt_raw_counts_df_20reads <- apply((filt_raw_counts_df), 1, 
                                    function(y) max(by(y, c_anno_df$condition, 
                                    function(x) sum(x>=reads_20))))

```

```{r}
filt_counts_df <- filt_raw_counts_df[filt_raw_counts_df_20reads >=case_5,]
row.names(r_anno_df) <- r_anno_df[,1]
filt_r_anno_df <- r_anno_df[rownames(filt_counts_df),]
dim(filt_counts_df) # number of genes filtered with genes > 20 reads and present in atleast 5 cases
dim(filt_r_anno_df)
# 16619 genes after the filtering.
```


```{r}
library(edgeR)
library(limma)
# create a DGE list object with the counts and annotation
counts <-  DGEList(counts=filt_raw_counts_df,
                group=factor((c_anno_df$condition)),
              genes=filt_r_anno_df[rownames(filt_raw_counts_df),])


# filter out the lowly expressed genes 
isexpr <- filterByExpr(counts, group = counts$samples$group)
table(isexpr) # number of genes highly expressed as true and lowly expressed as false 

# check if the genes with defined annotation, and recompute library sizes 
hasannot <- rowSums(is.na(counts$genes))== 0

# 15743 genes are expressed highly 
# 3037 genes are expressed lowly

```
```{r}
# keep the genes with the defined annotation 
counts <- counts[isexpr & hasannot, , keep.lib.sizes=FALSE]
# plot the librarry size of the samples 
barplot(counts$samples$lib.size*1e-6, names=colnames(counts), cex.axis=0.7, cex.names = 0.7,  ylab = "library size (millions)",las =2 )

```

```{r}
#apply TMM (trimmed means of M values between samples and the reference) normalisation to remove the composition biases:

counts <-  calcNormFactors(counts, method = 'TMM')
head(counts$samples)
# calculating the counts per million per sample for all the genes
counts_GeTMM <- as.data.frame(cpm(counts))
boxplot(counts_GeTMM, las =2, cex.axis=0.75, ylim=c(0,4000), ylab= "CPM")


```

```{r}
# after normalisation between samples
# library sizes (number of transcript counts per sample)
log_counts_GeTMM <- log2(counts_GeTMM +1)
boxplot(log_counts_GeTMM, las =2, cex.axis=0.75, ylab= "CPM")
```

```{r}
# estimate the dispersion between the case and the control using the Negative binomial dispersion as it regulates the gene outliers 

design <-  model.matrix(~0+group, data=counts$samples)
dispersion_counts<-  estimateDisp(counts, design, robust = TRUE)
#plotBCV(dispersion_counts)

```

```{r}
set.seed(111)
# differential gene expression with GLM quasi likelihood test 
# gene expression analysis 
# fit the model on the dispersion counts
fit <- glmQLFit(dispersion_counts, design, robust=TRUE)
#plotQLDisp(fit)
# the differentially expressed genes test with quasi likelihood F test
contrasts_cc <- makeContrasts("groupcase-groupcontrol", levels=design) 
de.genes <-  glmQLFTest(fit, contrast = contrasts_cc)
summary(decideTests(de.genes)) # no. of downregulated, upregulated and non significant genes

```

```{r}

library(ggplot2)
# volcano plot
# when cut-off of log2FC >=-1 and <=1 as non significant genes 
p1= plot(de.genes$table$logFC, -log(de.genes$table$PValue,base=10),xlab="log2FC", ylab="-log10Pvalue",
    col=ifelse(de.genes$table$logFC>=-1 & de.genes$table$logFC<=1,"grey70","red"), pch=20, frame.plot=TRUE, cex=0.8)

p1
# the red points are DE genes and the grey points are nonsignificant genes

```
```{r}
set.seed(111)
#select up- and down-regulated genes using a p-value cutoff of 0.01, a log fold change ratio >1.5 forup-regulated genes and < (-1.5) for down-regulated genes and a log CPM >1.
De_genes <-  as.data.frame(topTags(de.genes,n=12000,p.value = 0.01,sort.by = "logFC"))
De_genes$class <- "non-significant"
De_genes$class[which(De_genes$logCPM>1 & De_genes$logFC>1.5)] = "up"
De_genes$class[which(De_genes$logCPM>1 & De_genes$logFC<(-1.5) )] = "down"
```



```{r}
De_genes[De_genes$class== 'up', ] # 1118 genes upregulated 
De_genes[De_genes$class == 'down', ] #849 genes downregulated
```

```{r}
# remove non significant genes
De_genes <- De_genes[De_genes$class != 'non-significant',]

```


```{r}
# heatmap
color_matrix<-c_anno_df
color_matrix$color <- "red"
color_matrix$color[which(color_matrix$condition=="control")] <-"blue"
cols <- color_matrix$color

heatmap(as.matrix(counts_GeTMM[which(rownames(counts_GeTMM)%in%De_genes$ensembl_gene_id),]), 
                  ColSideColors = cols,cexCol = 0.6,margins = c(2,2),col=cols,cexRow = 0.6)
```

```{r}
### Use biomaRt to map Gene symbols, Entrez IDs and Ensembl gen IDs
ensembl <- useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl")
convert <- getBM(attributes=c("ensembl_gene_id","entrezgene_id","external_gene_name"),
                 filters=c("ensembl_gene_id"), 
                 values=De_genes$ensembl_gene_id,
                 mart = ensembl)

```

```{r}

De_genes_entrez <- merge(De_genes,convert,by.x="ensembl_gene_id",by.y="ensembl_gene_id")
De_genes_entrez <- De_genes_entrez[which(!is.na(De_genes_entrez$entrezgene_id)),]
De_genes_entrez <- De_genes_entrez[-which(duplicated(De_genes_entrez$entrezgene_id)),]
```


```{r}
DE_up <- De_genes_entrez[De_genes_entrez$class == "up", ]
DE_down <- De_genes_entrez[De_genes_entrez$class == "down", ]
```


```{r}
library(clusterProfiler)
#Gene set enrichment analysis
# for biological processes, for the upregulated genes
GSE_BP <- enrichGO(gene = DE_up$external_gene_name.x,
            OrgDb = 'org.Hs.eg.db',
            keyType = 'SYMBOL',
            ont = "BP" ,
            pAdjustMethod = "BH",
            pvalueCutoff = 0.05,
            qvalueCutoff = 0.05)
barplot(GSE_BP,showCategory=10) # top 10 enriched gene terms
# chromosome segregation is the most enriched biological process for the upregulated genes

```

```{r}

# for biological processes, for the downregulated genes
GSE_BP_down <- enrichGO(gene = DE_down$external_gene_name.x,
            OrgDb = 'org.Hs.eg.db',
            keyType = 'SYMBOL',
            ont = "BP" ,
            pAdjustMethod = "BH",
            pvalueCutoff = 0.05,
            qvalueCutoff = 0.05)
barplot(GSE_BP_down,showCategory=10) # top 10 enriched gene terms
# cilium movement is the most downregulated GO term for Biological process

```

```{r}
# GO for Molecular functions for the upregulated genes
GSE_MF <- enrichGO(gene = DE_up$external_gene_name.x,
OrgDb = 'org.Hs.eg.db',
keyType = 'SYMBOL',
ont = "MF" ,
pAdjustMethod = "BH",
pvalueCutoff = 0.05,
qvalueCutoff = 0.05)
barplot(GSE_MF,showCategory=10)

# extracellular matrix is the most enriched GO term for the Molecular Function
 
```

```{r}
# GO for Molecular functions for the downregulated genes
GSE_MF_down <- enrichGO(gene = DE_down$external_gene_name.x,
OrgDb = 'org.Hs.eg.db',
keyType = 'SYMBOL',
ont = "MF" ,
pAdjustMethod = "BH",
pvalueCutoff = 0.05,
qvalueCutoff = 0.05)
barplot(GSE_MF_down,showCategory=10)
# glycosaminoglycan binding is the most enriched GO term for molecular function

```



```{r}
# KEGG enrichment for Wiki pathways
ekegg <- enrichWP(gene = DE_up$entrezgene_id,
                    organism = "Homo sapiens",
          pvalueCutoff = 0.05,
  qvalueCutoff = 0.05)
as_tibble(ekegg, n=10)
```

```{r}
# WP for downregulated genes
ekegg_down <- enrichWP(gene = DE_down$entrezgene_id,
  organism = 'Homo sapiens',
  pvalueCutoff = 0.05,
  qvalueCutoff = 0.05)
as_tibble(ekegg_down, n=10)

```
```{r}
# . Use the pathview R package to visualize one pathway you find enriched using the 
#upregulated gene list.
library(pathview)
logFC <-DE_up$logFC
names(logFC) <- DE_up$entrezgene_id
#pathview(gene.data = logFC,  # the enrichKEGG did not work so no hsa id could be obtained for the pathway
#         pathway.id = "",
#         species = "human")
```

```{r}

library(MotifDb)
library(seqLogo)
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)

# How many motifs
 #length(MotifDb)

# See sources
#sort(table(values(MotifDb)$dataSource),decreasing=TRUE)

# See organisms
#sort(table(values(MotifDb)$organism),decreasing=TRUE)

# Which are the metadata
#colnames(values(MotifDb))

# Perform a query
res <- query(MotifDb,'hsapiens')
names(res)[1]
res[[1]]

```

```{r}
#Identify which transcription factors (TFs) have enriched scores in the promoters of all 
#up-regulated (or down-regulated if you prefer) genes. 
#a use a window of 500 nucleotides upstream each gene
downreg_genes <- DE_down$ensembl_gene_id # genes which are upregulated
promoter_seq <- getSequence(id = downreg_genes,  
                            type="ensembl_gene_id",
                            seqType="gene_flank",
                            upstream=500,  # 500 bp upstream
                            mart=ensembl) 
```


```{r}
data(PWMLogn.hg19.MotifDb.Hsap)
sequences <- lapply(promoter_seq$gene_flank,function(x) DNAString(x))
enriched_TFs <- motifEnrichment(sequences,PWMLogn.hg19.MotifDb.Hsap,score = "affinity")
report = groupReport(enriched_TFs)
report
plot(report[1:10], fontsize=7, id.fontsize=5) # top 10 most enriched TFs in downregulated genes
```
```{r}
# Select one among the top enriched TFs, compute the empirical distributions of scores 
#for all PWMs that you find in MotifDB for the selected TF and determine for all of 
#them the distribution (log2) threshold cutoff at 99.75%.
# Empirical distribution
tfs_motif = subset(MotifDb, organism=='Hsapiens' & geneSymbol=="PGAM2") # from the above enriched TFs the most
# enriched
PWM = toPWM(as.list(tfs_motif))
ecdf = motifEcdf(PWM,organism = "hg19",quick=TRUE)
thresholds = lapply(ecdf,function(x) log2(quantile(x,0.9975))) # 
thresholds
# the most enriched TF has only one motif Hsapiens-hPDI-PGAM2 with a score of 6.004127 at threshold cutoff at 99.75%.

```
8. Identify which up-regulated (or down-regulated depending on the choice you made 
at point 7) genes have a region in their promoter (defined as previously) with binding 
scores above the computed thresholds for any of the previously selected PWMs. 
a Use pattern matching as done during the course


```{r}
# Pattern matching
names(PWM) = sapply(names(PWM),function(x) strsplit(x,"-")[[1]][3])
cutoff_99.75 = c()
scores = motifScores(sequences,PWM,raw.score=FALSE,cutoff=unlist(thresholds), verbose = FALSE)
cutoff_99.75 = c(cutoff_99.75,length(which(apply(scores,1,sum)>0))/dim(DE_up)[1])
cutoff_99.75
# around 63% of all the genes contain this motif in their promoter region

```
```{r}
# the downregulated genes checked in STRING db
writeLines(DE_down$ensembl_gene_id, "C:/Users/surya/Downloads/Bioinformatic_resources/project_files-20230617/downregulated_genes.tsv")

```
9. Use STRING database to find PPI interactions among differentially expressed genes 
and export the network in TSV format. 

```{r}
# using STRING db for protein protein interactions
# for the downregulated genes
links <- as.data.frame(read.delim("C:/Users/surya/Downloads/Bioinformatic_resources/project_files-20230617/string_interactions_short.tsv"))
nodes <-
getBM(attributes=c("external_gene_name","ensembl_gene_id","description","gene_biotype","start_position","end_position","chromosome_name","strand"),
filters=c("ensembl_gene_id"),
values=DE_down[,1], # down regulated genes
mart = ensembl)
nodes = unique(nodes[,c(1,3:5)])
geneids = unique(links$X.node1[which(!links$X.node1%in%nodes$external_gene_name)])
geneids = unique(c(geneids,unique(links$node2[which(!links$node2%in%nodes$external_gene_name)])))
length(geneids)
# 61 genes are not linked to the PPI network of the upregulated genes 
# so, removing these genes
links <- links[which(!links$X.node1%in%geneids),]
links <- links[which(!links$node2%in%geneids),]

```
10. Import the network in R and using igraph package and identify and plot the largest 
connected component. 


```{r}
library(igraph)
net <- graph_from_data_frame(d=links,vertices=nodes,directed=FALSE)
class(net)

plot(net,
edge.width=5,
vertex.color="lightgreen",
vertex.size=10,
vertex.frame.color="darkgray",
vertex.label.color="black",
vertex.label.cex=0.3,
edge.curved=0.1)
# detection of the communities 
ceb <- cluster_edge_betweenness(net)


```
```{r}
plot(ceb, net,
vertex.size=10,
edge.width=5,
vertex.label.color="black",
vertex.label.cex=0.5,
edge.curved=0.1)
```
```{r}
component_c <- components(net, mode = c("strong"))
print(component_c$csize)
print(table(component_c$csize))
print(max(component_c$csize)) # 681 genes out of total up regulated genes are in the largest connected network in the network of all the upregulated genes
# largest connected network
netlargest.c = induced_subgraph(net,V(net)[which(component_c$membership==1)])
plot(netlargest.c,
edge.width=5,
vertex.color="lightgreen",
vertex.size=10,
vertex.frame.color="darkgray",
vertex.label.color="black",
vertex.label.cex=0.3,
edge.curved=0.1)

```


